stock_cancel/models/stock.py

```diff
@@ -46,16 +47,32 @@ def action_revert_done(self):
                     % (picking.name))
             if picking.invoice_id:
                 raise exceptions.Warning(
-                    _('Picking %s has invoices!') % (picking.name))
+                    _('Picking %s has been invoiced') % (picking.name))
+            for move in picking.move_lines:
+                quants = move.quant_ids.filtered(
+                    lambda q: q.location_id == move.location_dest_id)
+                if quants.filtered(lambda q: q.qty != move.product_uom_qty):
```

Comment:
What if a single move touched several quants? You'll refuse to undo it won't you?